src/capture/osx/appkit.rs

```diff
@@ -0,0 +1,198 @@
+use super::types::*;
+use crate::prelude::*;
+
+use objc::{
+    class, msg_send,
+    runtime::Object,
+    sel, sel_impl,
+};
+use std::ffi::CStr;
+use sysinfo::{
+    Pid, PidExt, ProcessExt, System, SystemExt,
+};
+
+pub struct OSXCapturer {
+    os_info: util::OsInfo
+}
+
+impl OSXCapturer {
+    pub fn init() -> OSXCapturer {
+        OSXCapturer{
+            os_info: util::get_os_info()
+        }
+    }
+}
+
+impl Capturer for OSXCapturer {
+    fn capture(&mut self) -> anyhow::Result<EventData> {
+       Ok(EventData::osx_v1(OSXEventData {
+            os_info: self.os_info.clone(),
+            focused_window: get_frontmost_app_pid(),
+            windows: get_running_apps(),
+            duration_since_user_input: user_idle::UserIdle::get_time()
+                .map(|e| e.duration())
+                .map_err(|e| anyhow::Error::msg(e))
+                .context("Couldn't get duration since user input")
+                .unwrap_or_else(|e| {
+                    log::warn!("{}", e);
+                    Duration::ZERO
+                })
+       }))
+    }
+}
+
+/// Frees any Objects
+unsafe fn release(object: *mut Object) {
+    let _: () = msg_send![object, release];
+}
+
+/// Turns an
+/// [NSString](https://developer.apple.com/documentation/foundation/nsstring?language=objc) Object into a `&str`.
+unsafe fn ns_string_to_string(ns_string: *mut Object) -> Option<String> {
+    // Get length of name
+    let string_size: usize = msg_send![ns_string, lengthOfBytesUsingEncoding: 4];
+
+    // Allocate length of name + 1 (for null terminator)
+    let char_ptr = libc::malloc(string_size + 1);
+
+    // Copy the string into the allocated memory
+    // encoding: 4 is for specifying that the string has UTF-8 encoding
+    let res: bool = msg_send![ns_string, getCString:char_ptr maxLength:string_size + 1 encoding:4];
+
+    release(ns_string);
+    
+    if !res {
+        libc::free(char_ptr);
+        return None;
+    }
+
+    let string = CStr::from_ptr(char_ptr as *const i8)
+        .to_str()
+        .unwrap()
+        .to_owned();
+
+    libc::free(char_ptr);
+
+    Some(string)
+}
+
+pub fn get_frontmost_app_pid() -> Option<i32> {
```

Comment:
usually, a process can have multiple windows. So using just the PID is probably not enough to identify a window?

Reply:
[frontmostApplication](https://developer.apple.com/documentation/appkit/nsworkspace/1532097-frontmostapplication?language=objc) is always pointing to the parent process from my observations, since it's supposed to return the frontmost application and not frontmost window, also for example when you spawn a new window in Chrome and you focus on it, then run the function it returns the same **pid** as the initial window.

Reply:
but we explicitly need the window information for the foremost window, e.g. if multiple chrome windows are open it needs to get the title / url of the window that's in the foreground

Reply:
I've just realised that it doesn't, it gets the application name, not the window name. I'll fix that.

Reply:
I've added proper querying of windows you can check the [commit](https://github.com/phiresky/timetrackrs/pull/14/commits/6f03985d8ca55c41041ca1508b95fc04a5596ee3).